kernel/mmanutils.c

```diff
@@ -849,33 +943,40 @@ void myst_mman_close_notify(int fd)
     /* if file is open for write */
     if (flags & (O_RDWR | O_WRONLY))
     {
-        myst_spin_lock(&_msync_mappings_lock);
+        myst_rspin_lock(&_mman.lock);
+        uint8_t* addr = (uint8_t*)_mman.map;
+        size_t length = ((uint8_t*)_mman.end) - addr;
+        myst_rspin_unlock(&_mman.lock);
+
+        vectors_t v = _get_vectors();
+        size_t index = _get_page_index(addr, length);
+        myst_assume(index >= 0);
+
+        myst_rspin_lock(&_mman.lock);
         {
-            msync_mapping_t* p = _msync_mappings;
-            msync_mapping_t* prev = NULL;
+            const size_t count = length / PAGE_SIZE;
+            myst_fdmapping_t* p = &v.fdmappings[index];
+            const myst_fdmapping_t* end = p + count;
+            size_t bytes_remaining = count * sizeof(myst_fdmapping_t);
 
-            while (p)
+            while (p < end)
             {
-                msync_mapping_t* next = p->next;
+                // Efficiently skip over zero-valued bytes. In-use fdmappings
+                // begin with a non-zero byte.
+                if ((p = myst_memcchr(p, '\0', bytes_remaining)) == NULL)
+                    break;
 
-                if (p->fd == fd)
+                if (p->used == MYST_FDMAPPING_USED && p->fd == fd)
```

Comment:
Clearing the fd mapping might conflict with writing back modifications at `munmap()` time for MAP_SHARED.

Reply:
This also affects reporting of pathname information in /proc/[pid]/maps.

Reply:
``mmap`` does not currently support ``MAP_SHARED``. This is something to keep in mind when we implemented shared file mappings.

Reply:
Should the mapping be preserved after the fd has already been closed? In any case, it is not something we supported before so we can create an issue if necessary.

Reply:
I think we supported it because /maps relied on process mappings which were released on process shutdown. But agree we can track it in an issue, should not be PR blocking.